enclave/src/openfinex/openfinex_client.rs

```diff
@@ -0,0 +1,382 @@
+            ws_version,
+            connection,
+            authorization,
+            host
+        );
+        debug!("Sending http request: {}", data);
+        if let Err(e) = self.socket.write(data.as_bytes()){
+            error!("Could not handshake with openfinex server: {}", e);
+        };
+    }
+
+    fn wants_write(&self) -> bool {
+        !self.sendable_plaintext.is_empty()
+    }
+
+    // we always want to read except when we already received some text
+    // we want to process
+    fn wants_read(&self) -> bool {
+        self.received_plaintext.is_empty()
+    }
+
+    /* fn parse_http_response(&mut self, buffer: &[u8]) -> c_int {
+        // split header and body
+        let mut index = 0;
+        let mut _prev_byte = 0;
+        let new_line = 0;
+        // Header - Body seperator: [13, 10, 13, 10] = \r\n\r\n
+        // start with header
+        for byte in buffer.iter() {
+            if *byte == 129 {
+                break;
+            } else if *byte == 0 {
+                // EOF
+                break;
+            }
+            index += 1;
+            _prev_byte = *byte;
+        }
+        let header = String::from_utf8_lossy(&buffer[0..index]);
+        debug!("Header: {}", header);
+        // get body
+        let body = String::new();
+        // TODO if necessary
+        debug!("Body: {}", body);
+        1
+    } */
+
+    /// We're ready to do a read.
+    fn do_read(&mut self) -> io::Result<usize> {
+        //FIXME: maybe we can use up buffer read here?
+        let mut buffer = [0 as u8; 512]; // Dummy buffer. will not be necessary with tls client
+        let rc = self.socket.read(&mut buffer);
+        if rc.is_err() {
+            error!("TLS read error: {:?}", rc);
+            //self.closing = true;
+            return rc;
+        }
+
+        // If we're ready but there's no data: EOF.
+        if rc.unwrap() == 0 {
+            error!("EOF");
+            //self.closing = true;
+            //self.clean_closure = true;
+            return Ok(0);
+        }
+
+        if buffer[0] == 72 {
+            // parse http response
+            error!(
+                "Received Http response: {}",
+                String::from_utf8_lossy(&buffer)
+            ); //FIXME: replace with debug
+               // handshake successful, send a subscription:
+            self.subscribe_matches();
+        } else {
+            // direct tcp message
+            if let Some(message) = client_utils::read_tcp_buffer(buffer.to_vec()) {
+                match message.opcode {
+                    Opcode::PingOp => {
+                        self.send_pong();
+                    }
+                    Opcode::TextOp => {
+                        debug!("received plaintext : {:?}", message.payload);
+                        self.response_handler.handle_text_op(message.payload);
+                    }
+                    _ => error!("received unexpected op: {:?}", message.opcode),
```

Comment:
We only accept text message (UFT8 encoded) or Pings from the server.. all other websocket calls will be ingored.